pkg/agent/protocol/grpc/proxy.go

```diff
@@ -77,40 +55,40 @@ func NewProxy(c ProxyConfig, d Disruption) (protocol.Proxy, error) {
 		return nil, fmt.Errorf("status code cannot be 0 (OK)")
 	}
 
-	return &proxy{
-		disruption: d,
-		config:     c,
-		metrics: protocol.NewMetricMap(
-			protocol.MetricRequests,
-			protocol.MetricRequestsExcluded,
-			protocol.MetricRequestsDisrupted,
-		),
-	}, nil
-}
-
-// Start starts the execution of the proxy
-func (p *proxy) Start() error {
-	// should receive the context as a parameter of the Start function
+	ctx, cancel := context.WithCancel(context.Background())
 	conn, err := grpc.DialContext(
-		context.Background(),
-		p.config.UpstreamAddress,
+		ctx,
+		upstreamAddress,
 		grpc.WithInsecure(),
```

Comment:
This code, which I believe starts a connection with the upstream GRPC server, was previously on `Start` while now is on the constructor. Right now I cannot think about any reasons this would make a difference, but I'm flagging it just in case.

Reply:
Could you elaborate on the rationale for moving it to the constructor, given that we already have the start method?

Reply:
Absolutely. For GRPC in particular, the handler itself needs to be supplied the connection to the upstream:

https://github.com/grafana/xk6-disruptor/blob/87377caa4a8affc5a7163e8dcfe24a50d795fd0f/pkg/agent/protocol/grpc/handler.go#L28-L28

Following that chain, the server needs the handler when it is being built, and the proxy needs the server when it is being built as well. I'm not extremely familiar with the GRPC library, but I couldn't find a way to pass the handler to the server after it has been built. As for passing the client to the handler, that is something we can do if we refactor it, something like:

```go
func NewProxy(listener net.Listener, upstreamAddress string, d Disruption) (protocol.Proxy, error) {
	// ...
	handler := NewHandler(d, metrics)
	// ...
}

func (p *proxy) Start() {
	// Create conn logic moved here.
	handler.useConn(conn)

	err := p.srv.Serve(p.listener)
	if err != nil {
		return fmt.Errorf("proxy terminated with error: %w", err)
	}

	return nil
}
```

I think this is doable, with the tradeoff of having a mutable handler that is not ready to use after being built. I think APIs designed this way are more bug prone as it's easy to forget that the returned object needs further initialization.

Another alternative could be to keep the handler as it is, and instead change the proxy interface from
```go
type Proxy interface {
	Start() error
	Stop() error
}
```
to
```go
type Proxy interface {
	Start(l net.Listener) error
	Stop() error
}
```

But I think we were not convinced by this, as we thought that maybe not all proxies will need a listener.

To summarize:
- We can leave it as it is, where we open the upstream connection before `Start`ing the proxy
- We can make the handler mutable and store some things in the proxy to defer connection opening to `Start`, introducing complexity
- We can change the `Proxy` interface so `Start` receives a listener.

LMK if this makes everything clearer. If you have alternatives to those, I'm all ears as I'm not a huge fan of either.

Reply:
Thanks for the detailed explanation. What I don't understand is why we need to create the server in the proxy constructor instead of in the start method as it is now.  It seems to me that this is the source of this chain you mention, but I don't think that is necessary or even convenient, unless I'm missing something.

Reply:
Doing so would not fix the race condition that triggers #319, where `Start` can initialize the server into the proxy struct at the same time `Stop` tries to access it.